core/libp2p/muxer/yamux/yamux.cpp

```diff
@@ -0,0 +1,521 @@
+          "connection error: read less bytes, than expected in header");
+      // terminate the Yamux?
+    }
+
+    if (!processHeader()) {
+      startReadingHeader();
+    }
+  }
+
+  void Yamux::readingDataCompleted(const std::error_code &ec, size_t n,
+                                   StreamParameters &stream) {
+    if (ec) {
+      logger_->error("cannot read from the connection: {}", ec.value());
+      // terminate Yamux?
+    }
+
+    // if there's a callback, which awaits for the message from this stream,
+    // call it; buffer the message otherwise
+    std::vector<uint8_t> bytes(n);
+    boost::asio::buffer_copy(boost::asio::buffer(bytes), read_buffer_.data());
+    read_buffer_.consume(n);
+    kagome::common::Buffer msg{std::move(bytes)};
+    if (!stream.completion_handlers_.empty()) {
+      stream.completion_handlers_.front()(std::move(msg));
+      stream.completion_handlers_.pop();
+    } else {
+      stream.buffered_messages_.push(std::move(msg));
+    }
+
+    startReadingHeader();
+  }
+
+  void Yamux::write(const common::NetworkMessage &msg,
+                    stream::Stream::ErrorCodeCallback cb) {
+    outcoming_messages_.push({msg, std::move(cb)});
+    if (!is_writing_) {
+      startWriting();
+    }
+  }
+
+  void Yamux::startWriting() {
+    if (!is_active_) {
+      return;
+    }
+    if (connection_->isClosed()) {
+      closeYamux();
+      return;
+    }
+    is_writing_ = true;
+
+    if (!outcoming_messages_.empty()) {
+      auto msg_and_callback = outcoming_messages_.front();
+      outcoming_messages_.pop();
+
+      write_buffer_ = std::move(msg_and_callback.first);
+      connection_->asyncWrite(
+          boost::asio::const_buffer{write_buffer_.toBytes(),
+                                    write_buffer_.size()},
+          [t = shared_from_this(), cb = std::move(msg_and_callback.second)](
+              const std::error_code &ec, size_t n) mutable {
+            t->writingCompleted(ec, n, cb);
+          });
+    } else {
+      is_writing_ = false;
+    }
+  }
+
+  void Yamux::writingCompleted(
+      const std::error_code &ec, size_t n,
+      const stream::Stream::ErrorCodeCallback &error_callback) {
+    error_callback(ec, n);
+    startWriting();
+  }
+
+  void Yamux::stop() {
+    is_active_ = false;
+  }
+
+  outcome::result<std::unique_ptr<stream::Stream>> Yamux::newStream() {
+    if (!is_active_) {
+      return YamuxErrorStream::YAMUX_IS_CLOSED;
+    }
+
+    auto stream_id = getNewStreamId();
+    write(newStreamMsg(stream_id),
+          [t = shared_from_this(), stream_id](auto &&ec, auto &&) {
+            if (ec) {
+              t->logger_->error(
+                  "could not write new stream message for stream_id {} with "
+                  "error {}",
+                  stream_id, ec.value());
+            }
+          });
+    streams_.insert(
+        std::make_pair(stream_id,
+                       std::make_shared<StreamParameters>(StreamParameters{
```

Comment:
why do you store plain struct StreamParameters on heap?

Reply:
Well, initial thought was that this struct could be exposed, but now it's useless, so yeah, going to turn it back from the shared_ptr